### PR TITLE
ISPN-16803 Move client modules to have the new client be the default

### PR DESCRIPTION
**Pull request by wburns**
https://issues.redhat.com/browse/ISPN-16803

### CODE REVIEW

**Comment by wburns** [comment]
Still can't reproduce the issue at https://ci.infinispan.org/job/Infinispan/job/PR-13126/8/testReport/org.infinispan.server.persistence/ManagedConnectionOperations/testTwoCachesSameCacheStore_Database__2_/

However, a blocking issue has been identified, which appears to just be an issue with needing a blockhound exclusion:
https://ci.infinispan.org/job/Infinispan/job/PR-13126/9/artifact/persistence/remote/target/surefire-reports/TEST-org.infinispan.persistence.remote.upgrade.HotRodUpgradeContainersEncodingsTest%5BHEAP%5D-Blocking.xml

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/hash/ConsistentHashTest.java [hunk 1]
```diff
@@ -46,9 +41,10 @@ public void testKeysMapToCorrectSegment() {
 
       for (int i = 0; i < 100; i++) {
          Object keyValue = kv(i);
-         ((ControlledChannelFactory) rcm.getChannelFactory()).useOnFetch((server, op) -> {
-            requests.put(keyValue, server);
-         });
+         // TODO: change this to something new
```
Comment (jabolina) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/hash/ConsistentHashTest.java:44]: Seems this test isn't doing anything.

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/retry/ClientListenerRetryTest.java [hunk 2]
```diff
@@ -60,11 +41,12 @@ public void testConnectionDrop() {
 
       assertListenerActive(remoteCache, listener);
 
-      failureInducingCodec.induceFailure();
+      // TODO: need to update to throw TransportException
+//      failureInducingCodec.induceFailure();
```
Comment (jabolina) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/retry/ClientListenerRetryTest.java:45]: This one seems is not testing a connection drop. IIRC it was already flaky :/

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/retry/DistributionRetryTest.java [hunk 3]
```diff
@@ -199,7 +183,7 @@ protected String parameters() {
    public Object[] factory() {
       return new Object[] {
             new DistributionRetryTest().withRetries(0),
-            new DistributionRetryTest().withRetries(10),
+//            new DistributionRetryTest().withRetries(10),
       };
```
Comment (jabolina) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/retry/DistributionRetryTest.java:187]: Should uncomment? This class was also flaky.

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/RemoteCache.java [hunk 4]
```diff
@@ -10,11 +10,11 @@
 import org.infinispan.client.hotrod.jmx.RemoteCacheClientStatisticsMXBean;
 import org.infinispan.commons.api.BasicCache;
 import org.infinispan.commons.api.TransactionalCache;
-import org.infinispan.commons.api.query.Query;
 import org.infinispan.commons.util.CloseableIterator;
 import org.infinispan.commons.util.CloseableIteratorCollection;
 import org.infinispan.commons.util.CloseableIteratorSet;
 import org.infinispan.commons.util.IntSet;
+import org.infinispan.query.dsl.Query;
 import org.reactivestreams.Publisher;
```
Comment (jabolina) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/RemoteCache.java:18]: I guess this change in the interface won't break compatibility somehow?

**Review by jabolina — COMMENTED** [review]
Done with the first commit. Just some minor points because of some tests, but I guess they were this way already in the client-new module.

Notice that a new test (`RemoteIndexFieldNameTest`) was integrated into the client-new last week and needs to be moved.

File: distribution/pom.xml [hunk 5]
```diff
@@ -22,11 +22,6 @@
          <artifactId>infinispan-query</artifactId>
       </dependency>
 
-      <dependency>
-         <groupId>org.infinispan</groupId>
-         <artifactId>infinispan-client-hotrod</artifactId>
-      </dependency>
-
```
Comment (jabolina) [inline distribution/pom.xml:29]: Is this right?

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/hash/ConsistentHashTest.java (hunk 1 above)
Comment (wburns) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/hash/ConsistentHashTest.java:44]: :+1: Let me fix this up, missed it.

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/retry/DistributionRetryTest.java (hunk 3 above)
Comment (wburns) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/retry/DistributionRetryTest.java:187]: Yeah, just adding back in.

**Review by theashiot — COMMENTED** [review]
I've added a few minor suggestions. Looks good otherwise.

best,

ashwin

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/hash/ConsistentHashTest.java (hunk 1 above)
Comment (wburns) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/hash/ConsistentHashTest.java:44]: Talking with @jabolina he noted we have https://issues.redhat.com/browse/ISPN-8098 which should be the way this is implemented.

File: documentation/src/main/asciidoc/topics/upgrading.adoc [hunk 6]
```diff
@@ -2,6 +2,19 @@
...
+The prior client used a connection pool to handle concurrent requests, moved to a new module `hotrod-client-legacy`.
+We now have a new client that instead only keeps a single connection per server it is connected to and supports
+pipelining by batching concurrent requests into a single flush operation.
+The new approach requires less system resources (file descriptors/threads) and also uses less CPU to submit the same task.
+
+The new client is effectively a drop in replacement. The only exceptions were if you were referencing some internal classes
+such as the ChannelFactory (non public API), which have been removed since they related to the underlying channel.
+
+The connection pool configuration is deprecated and no longer used at all and statistics related to the connection pool are
+deprecated and return -1 as they no longer are applicable.
```
[the remark this answers is not in the document]
Comment (wburns) [reply documentation/src/main/asciidoc/topics/upgrading.adoc:16]: I actually need to remove this as the stats were only ever on main and not in any released version.

File: distribution/pom.xml (hunk 5 above)
Comment (wburns) [reply distribution/pom.xml:29]: No.. odd. I will fix it.

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/RemoteCache.java (hunk 4 above)
Comment (wburns) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/RemoteCache.java:18]: Let me look at this closer.

Comment (wburns) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/RemoteCache.java:18]: Ah, it was just a conflict https://github.com/infinispan/infinispan/commit/54a9574e5c597ba69af8382c89ea2b2baa3cd53c. Reverting it.

File: distribution/pom.xml (hunk 5 above)
Comment (wburns) [reply distribution/pom.xml:29]: Oh, actually, yes it is correct. The cachestore-remote already includes this dependency.

File: documentation/src/main/asciidoc/topics/con_hotrod.adoc [hunk 7]
```diff
@@ -8,3 +8,9 @@ using different strategies.
 and automatically switch to available nodes.
 * Efficient data location. Hot Rod clients can find key owners and make
 requests directly to those nodes, which reduces latency.
+
+[NOTE]
+====
+The client only supports Hot Rod protocol version 3.0 and newer.
+The server supports Hot Rod protocol version 2.0 and newer.
```
[the remark this answers is not in the document]
Comment (wburns) [reply documentation/src/main/asciidoc/topics/con_hotrod.adoc:15]: Sure

**Comment by wburns** [comment]
Addressed all the comments, ran CI a few extra times to get some more runs in as well.

**Comment by wburns** [comment]
Green run and then another that had an OOM with some eclipse classes that seems unrelated. Reran CI a couple more times to see.

**Comment by tristantarrant** [comment]
@wburns unfortunately we now need a rebase because of the spatial query merge

**Comment by wburns** [comment]
Rebased.. the run that just failed was because when I rebased the last commit was lost somehow.. fixed so it should be good again.

**Comment by tristantarrant** [comment]
Thanks for the massive effort :)

**Pull request merged**